byzcoin/service.go

```diff
@@ -491,6 +581,167 @@ func (s *Service) createNewBlock(scID skipchain.SkipBlockID, r *onet.Roster, tx
 	return ssbReply.Latest, nil
 }
 
+// downloadDB downloads the full database over the network from a remote block.
+// It does so by copying the bboltDB database entry by entry over the network,
+// and recreating it on the remote side.
+// sb is a block in the byzcoin instance that we want
+// to download.
+func (s *Service) downloadDB(sb *skipchain.SkipBlock) error {
+	log.Lvlf2("%s: downloading DB", s.ServerIdentity())
+	idStr := fmt.Sprintf("%x", sb.SkipChainID())
+
+	// Loop over all nodes that are not the leader and
+	// not subleaders, to avoid overloading those nodes.
+	nodes := len(sb.Roster.List)
+	subLeaders := int(math.Ceil(math.Pow(float64(nodes), 1./3.)))
+	for ri := 1 + subLeaders; ri < nodes; ri++ {
+
+		// First delete an existing stateTrie. There
+		// cannot be another write-access to the
+		// database because s.catchingUp == true.
+		_, err := s.getStateTrie(sb.SkipChainID())
+		if err == nil {
+			// Suppose we _do_ have a statetrie
+			db, stBucket := s.GetAdditionalBucket(sb.SkipChainID())
+			err := db.Update(func(tx *bolt.Tx) error {
+				return tx.DeleteBucket(stBucket)
+			})
+			if err != nil {
+				log.Fatal("Cannot delete existing trie while trying to download:", err)
+			}
+			s.storage.Lock()
+			delete(s.stateTries, idStr)
+			s.storage.Unlock()
+		}
+
+		// Then start downloading the stateTrie over the network.
+		// Create a roster with just the node we want to
+		// download from.
+		roster := onet.NewRoster(sb.Roster.List[ri : ri+1])
+		cl := NewClient(sb.SkipChainID(), *roster)
+		var db *bolt.DB
+		var bucketName []byte
+		var nonce uint64
+		for {
+			resp, err := cl.DownloadState(sb.SkipChainID(), nonce, catchupFetchDBEntries)
+			if db == nil {
+				s.storage.Mutex.Lock()
+				db, bucketName = s.GetAdditionalBucket([]byte(idStr))
+				s.storage.Mutex.Unlock()
+				nonce = resp.Nonce
+			}
+			if err != nil {
+				return errors.New("cannot download trie: " + err.Error())
+			}
+			// And store all entries in our local database.
+			err = db.Update(func(tx *bolt.Tx) error {
+				bucket := tx.Bucket(bucketName)
+				for _, kv := range resp.KeyValues {
+					err := bucket.Put(kv.Key, kv.Value)
+					if err != nil {
+						return err
+					}
+				}
+				return nil
+			})
+			if err != nil {
+				log.Fatal("Couldn't store entries:", err)
+			}
+			if len(resp.KeyValues) < catchupFetchDBEntries {
+				break
+			}
+		}
+
+		// Check the new trie is correct
+		st, err := loadStateTrie(db, bucketName)
+		if err != nil {
+			return errors.New("couldn't load state trie: " + err.Error())
+		}
+		if sb.Index != st.GetIndex() {
```

Comment:
The database holding the trie has some extra fields (called MetaData) that are also copied when copying the database. Just some lines above the new `stateTrie` is loaded from the database, so that we can have the new index.

Reply:
OK, I knew about the metadata, but I wasn't sure they were being sent as well.